File: JavaScript/983-Minimum Cost For Tickets.js
/**
 * @param {number[]} days
 * @param {number[]} costs
 * @return {number}
 */
var mincostTickets = function(days, costs) {
    const N = days.length;
    const dayPass = [1, 7, 30];
    let memo = new Map();

    let dp = (i) =>{
        if (i >= N) return 0;
        if ( memo.has(i) ){
            return memo.get(i);
        }
        
        let nxtIdx = i, res = Number.POSITIVE_INFINITY;
        for (let j=0; j<costs.length; j++){
            while (nxtIdx < N && days[nxtIdx] < (dayPass[j] + days[i]) ){
                nxtIdx++;
            }
            res = Math.min(res, costs[j] + dp(nxtIdx));
        }
        memo.set(i, res);
        return res;
    }
    
    return dp(0);
};

File: C++/983-Minimum Cost For Tickets.cpp
class Solution {
public:
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        int N = days.size();
        vector<int> dp(N+1, INT_MAX);
        dp[0] = 0;
        for (int i=1; i<=N; i++) {
            for (int j=0; j<3; j++) {
                if (j==0) {
                    dp[i] = min(dp[i], dp[i-1]+costs[j]);
                } else {
                    int day = j==1?7:30;
                    auto it = lower_bound(begin(days), end(days), days[i-1]-day+1);
                    if (it == days.begin()) {
                        dp[i] = min(dp[i], costs[j]);
                    } else {
                        int idx = it - begin(days);
                        dp[i] = min(dp[i], costs[j] + dp[idx]);
                    } 
                }   
            }
        }

        return dp.back();
    }
};